07_errors/nicholascross/main.go

```diff
@@ -0,0 +1,157 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+type Puppy struct {
+	ID     int64
+	Breed  string
+	Colour string
+	Value  float64
+}
+
+type Storer interface {
+	CreatePuppy(p Puppy) (bool, error)
+	RetrievePuppy(int64) (*Puppy, error)
+	UpdatePuppy(int64, Puppy) (bool, error)
+	DeletePuppy(int64) (bool, error)
+}
+
+type MapStore map[int64]Puppy
+
+const (
+	missingPup      = 0
+	invalidPupValue = 1
+)
+
+type PupError struct {
+	Message string
+	Code    int
+}
+
+func (e *PupError) Error() string {
+	return fmt.Sprintf("[%d] %s", e.Code, e.Message)
+}
+
+func (e *PupError) IsMissingPup() bool {
+	return e.Code == missingPup
+}
+
+func (e *PupError) InvalidPupValue() bool {
+	return e.Code == invalidPupValue
+}
+
+func missingPuppy(id int64) error {
+	return &PupError{fmt.Sprintf("Puppy not found: %d", id), missingPup}
+}
+
+func invalidPuppyValue(value float64) error {
+	return &PupError{fmt.Sprintf("Puppy value must be non negative: %f", value), invalidPupValue}
+}
+
+var out io.Writer = os.Stdout
+
+func main() {
+	p := Puppy{1, "Jack Russel Terrier", "white and brown", 550.0}
+	fmt.Fprintf(out, "%d - %s [%s]: %f", p.ID, p.Breed, p.Colour, p.Value)
+}
+
+func (ms MapStore) CreatePuppy(p Puppy) (bool, error) {
+	if p.Value < 0 {
+		return false, invalidPuppyValue(p.Value)
+	}
+
```

Comment:
what if puppy with id `44` exist and p.ID == `44`? I believe you should return `false, AlreadyExist(id)` or sth simillar

Reply:
Anyway, I don't like situation when you are using `id` from request. Someone can accidentally modify some data and it can be pain on the neck in case of requests from distributed sources.

No action Required - decision is up to you

Reply:
Good point, I was treating it as no-op.  I have made this an error instead.